Update prometheus docs

hendrikmakait opened the pull request:
This PR updates the Prometheus documentation with the latest metrics and their descriptions and slight styling changes.

- [ ] Closes #9648
- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

hendrikmakait commented:
@crusaderky: Thanks for the review, I'll make some additions to the documentation here. Some of the metrics are only available in some cases, not in others. I'll lay the details out here. Good catch on the mismatching (sometimes matching) names, I'll have to dig deeper into how the Prometheus collector deals with and adjusts terms.

hendrikmakait commented:
> * Missing in documentation:
>   
>   * `process_virtual_memory_bytes`
>   * `process_resident_memory_bytes`
>   * `process_start_time_seconds`
>   * `process_cpu_seconds_total`
>   * `process_open_fds`
>   * `process_max_fds`

I think we should drop non-Dask metrics from the docs altogether. These are collected by the `prometheus-client` package, and keeping these in sync will be a mess. I'll just drop a link to the package on GitHub.

crusaderky commented:
@hendrikmakait this is now blocking https://github.com/dask/distributed/pull/7368; do you have an ETA?

hendrikmakait commented:
> * Missing in documentation:
>   
>   * `process_virtual_memory_bytes`
>   * `process_resident_memory_bytes`
>   * `process_start_time_seconds`
>   * `process_cpu_seconds_total`
>   * `process_open_fds`
>   * `process_max_fds`

I have now dropped all non-Dask metrics from the documentation. 

> * `dask_scheduler_tasks_forgotten` is not in /metrics. However I found `dask_scheduler_tasks_forgotten_total`.
> * `dask_scheduler_tasks_suspicious` is not in /metrics. However I found the HELP and TYPE of `dask_scheduler_tasks_suspicious_total`, but not the actual metric.
> * `dask_scheduler_prefix_state_totals` is not in /metrics. However I found the HELP and TYPE of `dask_scheduler_prefix_state_totals_total`, but not the actual metric.

I have fixed the naming on these in dask/distributed#7374. A general guideline here is not to use a suffix like `_total` for `Counter`s. The `prometheus_client` will automatically add these. (I have also noticed that we can specify units explicitly instead of suffixing names, but that's future work.)

> * These metrics have HELP and TYPE in /metrics, but not the actual metric:
>   
>   * `dask_semaphore_max_leases`
>   * `dask_semaphore_active_leases`
>   * `dask_semaphore_pending_leases`
>   * `dask_semaphore_acquire_total`
>   * `dask_semaphore_release_total`
> * `dask_semaphore_average_pending_lease_time` is not in /metrics. However I found the HELP and TYPE of `dask_semaphore_average_pending_lease_time_s`, but not the actual metric.

The individual metrics are dynamically generated for every existing semaphore, i.e., if you don't have a semaphore, no actual metric will show up here, only documentation for the metric group. I have a suspicion that we may want to use labels instead, but that's out of scope. 

> * `dask_stealing_request_count_total` is in /metrics but missing in the documentation
> * `dask_stealing_request_cost_total`  is in /metrics but missing in the documentation

I have added those to the documentation, good catch!

> * `dask_worker_concurrent_fetch_requests` has HELP and TYPE in /metrics, but no actual metric, and is missing in the documentation

I have added it to the documentation. I also see an actual metric here, not sure why you did not.

> * These are in the documentation but missing in /metrics:
>   
>   * `dask_worker_tick_duration_median_seconds`
>   * `dask_worker_task_duration_median_seconds`
>   * `dask_worker_transfer_bandwidth_median_bytes`

These are only exposed if `crick` is installed. I have added a note about that to the documentation

hendrikmakait commented:
I have restructured the docs into a general "how-to" part as well as individual segments for related metrics. This should facilitate a split as mentioned in https://github.com/dask/dask/issues/9648#issuecomment-1339617153

hendrikmakait commented:
Blocked by https://github.com/dask/distributed/pull/7374

crusaderky commented:
Could you make this PR align to main instead of #7374 so that we can unblock a wealth of stuff?
As discussed offline, we should really move this list to the distributed repo - something I'd like to do right after this PR is merged.

hendrikmakait commented:
@crusaderky: Done. Notably, `dask_semaphore_average_pending_lease_time_s` and `dask_scheduler_prefix_state_totals_total` are a thing again.
